File: Python/CPU project/CPU 10 archive/CPU 10 - old/ALU.py
class ALU:
	def __init__(self,input_bus,output_bus,op_bus):
		self.input_bus = input_bus
		self.output_bus = output_bus
		self.op_bus = op_bus

		self.reg1 = 0
		self.reg2 = 0
		self.flags = 0

		self.ops = [
		self.Add,
		self.Sub,
		self.Mul,
		self.Div,
		self.Mod,
		self.AND,
		self.OR,
		self.XOR,
		self.NOT,
		self.NAND,
		self.NOR,
		self.XNOR,
		self.SHL,
		self.SHR,
		self.ADDc,
		self.SUBb
		]



	def set_reg_1(self):
		self.flags = 0
		self.reg1 = 0
		for byte in self.input_bus.data:
			self.reg1 <<= 8
			self.reg1 += byte

	def set_reg_2(self):
		self.reg2 = 0
		for byte in self.input_bus.data:
			self.reg2 <<= 8
			self.reg2 += byte

	def enable_reg_1(self):
		for i in xrange(4):
			self.output_bus.data[3-i] = self.reg1&255
			self.reg1 >>= 8

	def enable_reg_2(self):
		for i in xrange(4):
			self.output_bus.data[i] = self.reg2&255
			self.reg2 >>= 8

	def enable_flags(self):
		self.output_bus.data = [0,0,0,self.flags]
		self.flags = 0

	def op(self):
		self.ops[self.op_bus.data[0]]()

	def Add(self):
		self.reg1 = self.reg1+self.reg2
		if self.reg1>4294967295:
			self.reg1 &= 4294967295
			self.flags = 16 #carry

	def Sub(self):
		self.reg1 = self.reg1-self.reg2
		if self.reg1<0:
			self.reg1 &= 4294967295
			self.flags = 8 #borrow

	def Mul(self):
		self.reg1 = self.reg1*self.reg2
		self.reg2 = self.reg1 >>32 #upper bytes
		self.reg1 &= 4294967295 #lower bytes

	def Div(self):
		if self.reg2 == 0:
			self.reg1 = 4294967295
			self.flags = 4 #division by 0
		else:
			self.reg1 //= self.reg2


	def Mod(self):
		if self.reg2 == 0:
			self.reg1 = 0
			self.flags = 4 #division by 0
		else:
			self.reg1 %= self.reg2

	def AND(self):
		self.reg1 &= self.reg2

	def OR(self):
		self.reg1 |= self.reg2
	def XOR(self):
		self.reg1 ^= self.reg2
	def NOT(self):
		self.reg1 ^= 4294967295

	def NAND(self):
		self.reg1 = (self.reg1&self.reg2)^4294967295

	def NOR(self):
		self.reg1 = (self.reg1|self.reg2)^4294967295

	def XNOR(self):
		self.reg1 = (self.reg1^self.reg2)^4294967295

	def SHL(self):
		self.reg1 = (self.reg1<<self.reg2)&4294967295

	def SHR(self):
		self.reg1 = (self.reg1>>self.reg2)&4294967295

	def ADDc(self):
		self.reg1 = self.reg1+self.reg2+1
		if self.reg1>4294967295:
			self.reg1 &= 4294967295
			self.flags = 16 #carry
	def SUBb(self):
		self.reg1 = self.reg1-self.reg2-1
		if self.reg1<0:
			self.reg1 &= 4294967295
			self.flags = 8 #borrow

File: C/CPU 10/ALU.c
#include <stdio.h>


// __________________ ALU _____________________________
void ALU_op(int *to_return, unsigned int r1_value, unsigned int r2_value, unsigned int op){ //need to pass an array to the function
    unsigned int result = 0;
	unsigned int top_of_result = 0;
	unsigned long long_result = 0;
	int flags = 0;

	if (op == 0){  			//ADD
		result = r1_value + r2_value;
		if (result < r1_value){	//if there is a carry
			flags = 16;
		}
	}
	else if (op == 1){		//SUB
		result = r1_value - r2_value;
		if (result > r1_value){ 	//if there is a borrow
			flags = 8;
		}


	}
	else if (op == 2){		//MUl
		long_result = (long)r1_value * (long)r2_value;
		result = long_result & 4294967295;
		top_of_result = long_result>>32;
	}
	else if (op == 3){		//DIV
		if  (r2_value == 0){ 	//division by 0
			flags = 4;
			result = 4294967295;	
		}
		else {
			result = r1_value/r2_value;
		}
	}
	else if (op == 4){		//MOD
		if (r2_value == 0){
			flags = 4;
			result = 0;
		}
		else {
			result = r1_value % r2_value;
		}

	}
	else if (op == 5){		//AND
		result = r1_value & r2_value;
	}
	else if (op == 6){		//OR
		result = r1_value | r2_value;
	}
	else if (op == 7){		//XOR
		result = r1_value ^ r2_value;
	}
	else if (op == 8){ 		//NOT
		result = r1_value ^ 4294967295;
	}
	else if (op == 9){		//NAND
		result = (r1_value & r2_value) ^ 4294967295;
	}
	else if (op == 10){ 		//NOR
		result = (r1_value | r2_value) ^ 4294967295;
	}
	else if (op == 11){		//XNOR
		result = (r1_value ^ r2_value) ^ 4294967295;
	}
	else if (op == 12){		//SHL
		result = r1_value << r2_value;

	}
	else if (op == 13){		//SHR
		result = r1_value >> r2_value;

	}
	else if (op == 14){		//ADDc
		result = r1_value + r2_value + 1;
		if (result < r1_value){	//if there is a carry
			flags = 16;
		}
	}
	else if (op == 15){		//SUBb
		result = r1_value - 1 - r2_value;
		if (result > r1_value){ 	//if there is a borrow
			flags = 8;
		}
	}
	to_return[0] = result;
	to_return[1] = top_of_result;
	to_return[2] = flags;
}

int main(){
	int result[3];
	printf("working\n");
    ALU_op(result,15,16,1);
	printf("bottom of result = %u\n",result[0]);

	printf("top of result = %u\n",result[1]);

	printf("flags = %u\n",result[2]);
}
